settings: Add organization setting to show message moves only.

madrix01 opened the pull request:
Issue #21398

timabbott commented on zerver/views/message_edit.py:
```diff
@@ -93,13 +93,17 @@ def get_message_edit_history(
+        if user_profile.realm.show_message_edit_history == 2:
+            for history in raw_edit_history[:]:
+                if "prev_content" in history:
+                    raw_edit_history.remove(history)
```
Let's make this block a function, perhaps `visible_edit_history_for_message(message, realm)`; we can put it in `zerver/lib/message.py` and use it here as well as other code paths that need it.

madrix01 replied:
This is a good way to go, on it!

timabbott commented on zerver/models.py:
```diff
@@ -676,7 +685,7 @@ class Realm(models.Model):
     property_types: Dict[str, Union[type, Tuple[type, ...]]] = dict(
         add_custom_emoji_policy=int,
-        allow_edit_history=bool,
+        show_message_edit_history=int,
```
`message_edit_history_visibility` is probably a better name.

timabbott commented on zerver/models.py:
```diff
@@ -438,6 +438,15 @@ class Realm(models.Model):
+        SHOW_EDITED_AND_MOVED,
+        SHOW_MOVED_ONLY,
+        SHOW_NO_HISTORY,
+    ]
```
We have a convention for these names. This should be defined just before the setting these values are for, and be named e.g. `MESSAGE_EDIT_HISTORY_VISIBILITY_NONE`, `MESSAGE_EDIT_HISTORY_VISIBILITY_MOVES_ONLY`, `MESSAGE_EDIT_HISTORY_VISIBILITY_ALL`.

timabbott commented on zerver/lib/message.py:
```diff
@@ -212,9 +212,16 @@ def messages_for_ids(
+                if len(msg_dict["edit_history"]) == 0:
+                    del msg_dict["edit_history"]
+            elif show_message_edit_history == 3:
+                del msg_dict["edit_history"]
```
Ideally, we'd arrange that this call the same function that I suggested adding in the views code, so we're not implementing this logic twice. (That oftens leads to bugs).  You may need to fiddle with the interface to make it work well.

timabbott commented:
I posted a batch of comments on some details I noticed. We'll also want to document the API change with `**Changes**`; our new feature tutorial and OpenAPI docs should provide the resources on how to do so correctly (https://zulip.readthedocs.io/en/latest/tutorials/new-feature-tutorial.html). 

I think ideally this would be structured as two commits: A first commit that changes the API from a boolean to an integer (with no functional changes), and then a second that implements the new value. Can you rework it to that structure?

madrix01 commented:
@timabbott Yes, I'll start work on restructuring PR, I'll make commit tomorrow with change in model from `bool` to `int` and should it include name change too?

timabbott commented:
Yes, of course; the previous name only makes sense for a boolean property. It'll be the vast majority of this PR -- just leaving off the bit that implements the new functionality.

madrix01 commented:
@timabbott I have pushed the preparatory commit, but it's strange that the CI didn't run and is still showing the merge conflicts.

madrix01 commented:
I have updated the preparatory commit along with tests.

timabbott commented on templates/zerver/api/changelog.md:
```diff
@@ -20,8 +20,10 @@ format used by the Zulip server that they are interacting with.
+**Feature level 125**
+
+* Replacing the previous `allow_edit_history` boolean field to
+  `message_edit_history_visibility` integer field.
```
This should list the endpoints affected with links, like in other examples in this file. In general it's always worth looking for something similar to see the preferred style in a given part of the codebase,

timabbott commented on zerver/migrations/0389_message_edit_history_visibility.py:
```diff
@@ -0,0 +1,22 @@
+            name="message_edit_history_visibility",
+            field=models.PositiveSmallIntegerField(default=1),
+        ),
+    ]
```
This migration doesn't transfer the existing settings. See https://zulip.readthedocs.io/en/latest/subsystems/schema-migrations.html

madrix01 committed:
settings: Change `allow_edit_history` to integer field.

We replace previously `allow_edit_history` boolean field to
`message_edit_history_visibility` integer field in order to add
an intermediate setting to view only MOVED history. The functionality
of the mentioned feature will be incorporated in upcoming commit.

Fixes a part of zulip#21398

alya commented:
@timabbott There is a question for you on this one as to whether your feedback above has been addressed correctly.

madrix01 commented:
@timabbott @alya Does this needs to be completed, I can pick this up ??

alya commented:
Sure, this still seems helpful -- thanks!

saubhagya-patel commented:
@madrix01, are you still working on this one? If not, I would like to continue.
Thank you.

saubhagya-patel commented:
I will be picking this up.
